x-pack/plugins/enterprise_search/public/applications/app_search/components/synonyms/synonyms.tsx

```diff
@@ -5,23 +5,86 @@
  * 2.0.
  */
 
-import React from 'react';
+import React, { useEffect } from 'react';
 
-import { EuiPageHeader, EuiPageContentBody } from '@elastic/eui';
+import { useValues, useActions } from 'kea';
+
+import {
+  EuiPageHeader,
+  EuiButton,
+  EuiPageContentBody,
+  EuiSpacer,
+  EuiFlexGrid,
+  EuiFlexItem,
+  EuiPagination,
+} from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
 
 import { FlashMessages } from '../../../shared/flash_messages';
 import { SetAppSearchChrome as SetPageChrome } from '../../../shared/kibana_chrome';
+import { Loading } from '../../../shared/loading';
 import { getEngineBreadcrumbs } from '../engine';
 
+import { SynonymCard, EmptyState } from './components';
 import { SYNONYMS_TITLE } from './constants';
 
+import { SynonymsLogic } from './';
+
 export const Synonyms: React.FC = () => {
+  const { loadSynonyms, onPaginate } = useActions(SynonymsLogic);
+  const { synonymSets, meta, dataLoading } = useValues(SynonymsLogic);
+  const hasSynonyms = synonymSets.length > 0;
+
+  useEffect(() => {
+    loadSynonyms();
+  }, [meta.page.current]);
+
+  useEffect(() => {
+    // If users delete the only synonym set on the page, send them back to the previous page
+    if (!hasSynonyms && meta.page.current !== 1) {
+      onPaginate(meta.page.current - 1);
+    }
+  }, [synonymSets]);
+
+  if (dataLoading && !hasSynonyms) return <Loading />;
+
   return (
     <>
       <SetPageChrome trail={getEngineBreadcrumbs([SYNONYMS_TITLE])} />
-      <EuiPageHeader pageTitle={SYNONYMS_TITLE} />
+      <EuiPageHeader
+        pageTitle={SYNONYMS_TITLE}
+        rightSideItems={[
+          <EuiButton fill onClick={() => {} /* TODO */}>
+            {i18n.translate(
+              'xpack.enterpriseSearch.appSearch.engine.synonyms.createSynonymSetButtonLabel',
+              { defaultMessage: 'Create a synonym set' }
+            )}
+          </EuiButton>,
+        ]}
+      />
       <FlashMessages />
-      <EuiPageContentBody>TODO</EuiPageContentBody>
+      <EuiPageContentBody>
+        <EuiSpacer size="s" />
+        {hasSynonyms ? (
+          <>
+            <EuiFlexGrid columns={3}>
+              {synonymSets.map(({ id, synonyms }) => (
+                <EuiFlexItem key={id}>
+                  <SynonymCard id={id} synonyms={synonyms} />
+                </EuiFlexItem>
+              ))}
+            </EuiFlexGrid>
+            <EuiSpacer />
+            <EuiPagination
+              pageCount={meta.page.total_pages}
+              activePage={meta.page.current - 1}
+              onPageClick={(pageIndex) => onPaginate(pageIndex + 1)}
+            />
```

Comment:
Forgot one last comment here - I wasn't sure if I should extract this out into a shared helper similar to our [table pagination](https://github.com/elastic/kibana/tree/master/x-pack/plugins/enterprise_search/public/applications/shared/table_pagination) helpers... 🤔

Unfortunately it's not actually _super_ similar to EuiBasicTable's `pagination` prop so I couldn't reuse that but it's _somewhat_ similar which makes DRYing it out/refactoring it with table pagination in mind slightly tricky.

I think this is the only place in App Search we use an `EuiPagination` component by itself, so I didn't bother for now, but if y'all think it's worth it I can do so!

Reply:
I think I'll just leave it for now, if we add another `EuiPagination` app in the app somewhere we can DRY it out then!